### ISSUE TITLE
Weird intersection when the track is tangent to the ZCylinder

### ISSUE THREAD

**Issue by alephpiece** [mit-crpg/OpenMOC#456]
The intersection is not as expected when the track is tangent to the ZCylinder. I believe that the coordinates `(xcurr, ycurr)` should always be computed before updating length `interior`, and the out parameter `points` should never participate in computation.

This patch seems to work.

```
--- a/src/Surface.cpp
+++ b/src/Surface.cpp
@@ -1002,11 +1002,11 @@ int ZCylinder::intersection(Point* point, double azim, double polar, Point* poin

     /* There is one intersection (ie on the Surface) */
     else if (fabs(discr) < FLT_EPSILON) {
+      xcurr = -b / (2*a);
+      ycurr = y0 + m * (xcurr - x0);
       double interior = pow(ycurr - y0, 2.0) + pow(xcurr - x0, 2.0);
       if (interior < 0.0)
         interior = 0.0;
-      xcurr = -b / (2*a);
-      ycurr = y0 + m * (points[num].getX() - x0);
       zcurr = z0 + sqrt(interior) * tan(M_PI_2 - polar);
       points[num].setCoords(xcurr, ycurr, zcurr);
```

**Comment by GiudGiud** [comment]
Good catch! PR #457 should fix this

**Comment by GiudGiud** [comment]
Closed by #457
